### PR TITLE
KAFKA-7922: Return authorized operations in Metadata request response (KIP-430 Part-2)

### PR DESCRIPTION
**Pull request by omkreddy**
-  Use automatic RPC generation in Metadata Request/Response classes
-  https://cwiki.apache.org/confluence/display/KAFKA/KIP-430+-+Return+Authorized+Operations+in+Describe+Responses

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by omkreddy** [comment]
This PR is built on top of #6322. Will rebase the PR after we merge #6322.

File: core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala [hunk 1]
```diff
@@ -43,6 +44,7 @@ class MetadataRequestTest extends BaseRequestTest {
     doSetup(createOffsetsTopic = false)
   }
 
+  //FIX ME
   @Test
   def testClusterIdWithRequestVersion1() {
     val v1MetadataResponse = sendMetadataRequest(MetadataRequest.Builder.allTopics.build(1.toShort))
```
Comment (omkreddy) [inline core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala]: @cmccabe Currently automatic protocol generator doesn't support null value as default. For MetadataRequest we need to return null clusterId  for V1 requests. Any thoughts on handling this?

```
kafka.server.MetadataRequestTest > testClusterIdWithRequestVersion1 FAILED
    java.lang.AssertionError: v1 clusterId should be null expected null, but was:<>
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotNull(Assert.java:756)
        at org.junit.Assert.assertNull(Assert.java:738)
        at kafka.server.MetadataRequestTest.testClusterIdWithRequestVersion1(MetadataRequestTest.scala:51)
```
cc @rajinisivaram  @ijuma

File: clients/src/main/java/org/apache/kafka/clients/admin/TopicDescription.java [hunk 2]
```diff
@@ -87,9 +90,20 @@ public boolean isInternal() {
...
+     */
+    public Set<AclOperation>  authorizedOperations() {
+        return authorizedOperations;
+    }
+
     @Override
     public String toString() {
-        return "(name=" + name + ", internal=" + internal + ", partitions=" +
-            Utils.join(partitions, ",") + ")";
+        return "TopicDescription{" +
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/admin/TopicDescription.java]: nit: For consistency, perhaps this should be `(name=...)` without `TopicDescription{}`?

File: clients/src/main/java/org/apache/kafka/common/requests/MetadataRequest.java [hunk 3]
```diff
@@ -17,203 +17,164 @@
...
             default:
                 throw new IllegalArgumentException(String.format("Version %d is not valid. Valid versions for %s are 0 to %d",
                         versionId, this.getClass().getSimpleName(), ApiKeys.METADATA.latestVersion()));
         }
     }
 
     public boolean isAllTopics() {
-        return topics == null;
+        return (data.topics() == null) ||
+            (data.topics().size() == 0 && version == 0); //In version 0, an empty topic list indicates
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/requests/MetadataRequest.java]: nit: `.size() ==0` => `isEmpty()`?

File: clients/src/main/java/org/apache/kafka/common/requests/MetadataRequest.java [hunk 4]
```diff
@@ -17,203 +17,164 @@
...
+                                                         // "request metadata for all topics."
     }
 
     public List<String> topics() {
-        return topics;
+        if ((data.topics() == null) ||
+            (data.topics().size() == 0 && version == 0)) //In version 0, we return null for empty topic list
+            return null;
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/requests/MetadataRequest.java:160]: Could use the method above (`isAllTopics`)?

File: clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java [hunk 5]
```diff
@@ -58,239 +56,64 @@
...
+    public MetadataResponse(MetadataResponseData data) {
+        this.data = data;
     }
 
-    private final int throttleTimeMs;
-    private final Collection<Node> brokers;
-    private final Node controller;
-    private final List<TopicMetadata> topicMetadata;
-    private final String clusterId;
+    private List<TopicMetadata> topicMetaData() {
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java]: nit: inconsistent case: `topicMetadata()`?

File: clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java [hunk 6]
```diff
@@ -363,14 +186,15 @@ public int throttleTimeMs() {
         return invalidMetadataTopics;
     }
 
+
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java]: nit: unnecessary newline

**Review by rajinisivaram — COMMENTED** [review]
@omkreddy Thanks for the PR, looks good. Left a few minor comments. Will wait to see what @cmccabe says about null values in the protocol generator.

File: core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala (hunk 1 above)
Comment (cmccabe) [reply core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala]: Hi @omkreddy, that's a good point. I think we should allow null default values.  I have a PR here which implements that here: https://github.com/apache/kafka/pull/6387

**Comment by omkreddy** [comment]
retest this please

**Comment by omkreddy** [comment]
@rajinisivaram  Thanks for the review. Addressed the review comments and Updated the response protocol to support null default values for clusterId.

**Review by rajinisivaram — APPROVED** [review]
@omkreddy Thanks for the updates, LGTM

**Pull request closed without merging**